File: scala/P01.scala
// P01 Find the last element of a list.

// Predefined function
def f0[T](list: List[T]): T = list.last

// Tail recursion
@annotation.tailrec
def f1[T](list: List[T]): T = list match {
  case h :: Nil => h
  case _ :: t => f1(t)
  case _ => throw new NoSuchElementException("List is empty")
}

// Chained calls
def f2[T](list: List[T]): T = list.reverse.head

// Haskell-like function composition, point-free style (looks strange with Scala syntax and OO model)
def f3[T] = ((_: List[T]).head).compose((_: List[T]).reverse)

// Reversed function composition
def f4[T] = ((_: List[T]).reverse).andThen((_: List[T]).head)

// Java-like implementation with side effects
def f5[T](list: List[T]): T = {
  var result: Option[T] = None
  for (each <- list) result = Some(each)
  result.getOrElse(throw new NoSuchElementException("List is empty"))
}

// Folding
def f6[T](list: List[T]): T = if (list.nonEmpty) list.reduce((a, b) => b) else
  throw new NoSuchElementException("List is empty")

File: haskell/p01.hs
-- P01 Find the last element of a list.

-- Predefined function
f0 :: [a] -> a
f0 = last

-- Recursion
f1 :: [a] -> a
f1 [] = error "empty list"
f1 [x] = x
f1 (_:xs) = f1 xs

-- Function application
f2 :: [a] -> a
f2 xs = head $ reverse xs

-- Function composition
f3 :: [a] -> a
f3 = head . reverse

-- Folding
f4 :: [a] -> a
f4 = foldl (curry snd) (error "empty list")

f5 :: [a] -> a
f5 = foldr1 (flip const)

f6 :: [a] -> a
f6 = foldr1 (const id)

f7 :: [a] -> a
f7 = foldr1 (\_ y -> y)